terraso_backend/apps/soil_id/graphql/soil_id.py

```diff
@@ -0,0 +1,211 @@
+
+from apps.soil_id.graphql.soil_data import (
+    DepthDependentSoilDataNode,
+    DepthIntervalInput,
+    SoilDataNode,
+)
+
+
+class EcologicalSite(graphene.ObjectType):
+    """Information about an ecological site."""
+
+    name = graphene.String(required=True)
+    id = graphene.String(required=True)
+    url = graphene.String(required=True)
+
+
+class SoilSeries(graphene.ObjectType):
+    """Information about a soil series."""
+
+    name = graphene.String(required=True)
+    taxonomy_subgroup = graphene.String(required=True)
+    description = graphene.String(required=True)
+    full_description_url = graphene.String(required=True)
+    data_source = graphene.String(required=True)
+
+
+class LandCapabilityClass(graphene.ObjectType):
+    """Caveat: may want to update these fields to an enum at some point."""
+
+    capability_class = graphene.String(required=True)
+    sub_class = graphene.String(required=True)
+
+
+class SoilInfo(graphene.ObjectType):
+    """Provides information about soil at a particular location."""
+
+    soil_series = graphene.Field(SoilSeries, required=True)
+    ecological_site = graphene.Field(EcologicalSite, required=False)
+    land_capability_class = graphene.Field(LandCapabilityClass, required=True)
+
+
+class SoilMatch(graphene.ObjectType):
+    """The likelihood score and rank within the match group for a particular soil type."""
+
+    score = graphene.Float(required=True)
+    rank = graphene.Int(required=True)
+
+
+class LocationBasedSoilMatch(graphene.ObjectType):
+    """A soil match based solely on a coordinate pair."""
+
+    soil_info = graphene.Field(SoilInfo, required=True)
+    match = graphene.Field(SoilMatch, required=True)
+
+
+class LocationBasedSoilMatches(graphene.ObjectType):
+    """A ranked group of soil matches based solely on a coordinate pair."""
+
+    matches = graphene.Field(graphene.List(LocationBasedSoilMatch), required=True)
+
+
+class DataBasedSoilMatch(graphene.ObjectType):
+    """A soil match based on a coordinate pair and soil data."""
+
+    soil_info = graphene.Field(SoilInfo, required=True)
+    location_match = graphene.Field(SoilMatch, required=True)
+    data_match = graphene.Field(SoilMatch, required=True)
+    combined_match = graphene.Field(SoilMatch, required=True)
+
+
+class DataBasedSoilMatches(graphene.ObjectType):
+    """A ranked group of soil matches based on a coordinate pair and soil data."""
+
+    matches = graphene.Field(graphene.List(DataBasedSoilMatch), required=True)
+
+
+class SoilIDInputDepthDependentData(graphene.InputObjectType):
+    """Depth dependent data provided to the soil ID algorithm."""
+
+    depth_interval = graphene.Field(DepthIntervalInput, required=True)
+    texture = graphene.Field(DepthDependentSoilDataNode.texture_enum())
+    rock_fragment_volume = graphene.Field(DepthDependentSoilDataNode.rock_fragment_volume_enum())
+    color_hue = graphene.Float()
+    color_value = graphene.Float()
+    color_chroma = graphene.Float()
```

Comment:
Question: I'm slightly surprised these are represented as Floats, is that just how the soil color algorithm works?

Reply:
My understanding is the YR is basically an enum. It could be an integer.

Reply:
my understanding/thinking on this:

- hue gets split into 10 steps, being `R YR Y GY G BG B PB P RP`, and each step is split into 4 substeps, being `2.5 5 7.5 10`
- value and chroma when displayed are rounded to nearest soil color (or to nearest integer if they're far away from nearest soil color)
- if chroma is close to 0, instead the whole thing is displayed as `N $VALUE/`

we store them as floats because the only time we actually need them in munsell notation format is for user display, but all internal manipulation of the data is easier in numerical space, and we also avoid erasing information unnecessarily this way (e.g. it might be helpful for iterating on the color algorithm to see what the unrounded outputs are)

Reply:
Ok, I still don't entirely understand how to translate the floats into munsell notation, but as long as you know how to, I'm good with this :)

Reply:
there is [a function for that](https://github.com/techmatters/terraso-mobile-client/blob/ada733f533d008d1b02738aaa26cba949556be63/dev-client/src/screens/SoilScreen/ColorScreen/utils/munsellConversions.ts#L201-L215) and [a longer explanation](https://github.com/techmatters/terraso-mobile-client/blob/ada733f533d008d1b02738aaa26cba949556be63/dev-client/src/screens/SoilScreen/ColorScreen/utils/munsellConversions.ts#L18-L42) in the file `munsellConversions.ts`